Automattic/wp-calypso
# composite-checkout: Add redirect when payment complete

#### Changes proposed in this Pull Request

This PR modifies checkout to add an additional form status. Previously there was `loading`, `ready`, and `submitting`. This adds `complete`. It also removes `usePaymentComplete` from the public API. Instead of calling `onPaymentComplete` directly when a Payment Method object finishes, it now changes the form status to `complete`. This will automatically trigger the `onPaymentComplete` callback.

Using this technique, this PR redirects to a "complete" page when the purchase is successful.

It also modifies the "jump-to-step" behavior added in #38493 so that it only jumps when all previous steps are complete.

#### Testing instructions

First test the demo.

- Run `npm run composite-checkout-demo`.
- Complete the form using the Credit card payment method (use the Stripe test card of `4242424242424242`) and press "Pay".
- Verify that there is a short pause during which the button is disabled, then the page should redirect to a "complete" page.

Then test Calypso.

- Sandbox the store.
- Run calypso with `ENABLE_FEATURES=composite-checkout-wpcom npm start`
- Add a plan to your cart (not a domain) and visit checkout.
- Complete the form using the Credit card payment method (use the Stripe test card of `4242424242424242`) and press "Pay".
- Verify that there is a short pause during which the button is disabled, then the page should redirect to a "complete" page.
